core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright 2012 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import java.util.Map.Entry;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.util.internal.StringUtil;
+
+final class QueryStringEncoder {
+
+    // Forked from netty-4.1.43.
+    // https://github.com/netty/netty/blob/bd8cea644a07890f5bada18ddff0a849b58cd861/codec-http/src/main/java/io/netty/handler/codec/http/QueryStringEncoder.java
+
+    private static final byte WRITE_UTF_UNKNOWN = (byte) '?';
+    private static final char[] CHAR_MAP = "0123456789ABCDEF".toCharArray();
+
+    static void encodeParams(StringBuilder buf, QueryParamGetters params) {
+        for (Entry<String, String> e : params) {
+            encodeUtf8Component(buf, e.getKey());
+            final String value = e.getValue();
+            if (value != null) {
+                buf.append('=');
+                encodeUtf8Component(buf, value);
+            }
+            buf.append('&');
+        }
+
+        buf.setLength(buf.length() - 1);
+    }
+
+    /**
+     * Encodes a query component (name or value).
+     *
+     * @see ByteBufUtil#writeUtf8(ByteBuf, CharSequence, int, int)
+     */
+    private static void encodeUtf8Component(StringBuilder uriBuilder, String s) {
```

Comment:
It'd be good to compare this to guava's `URLEscapers` or JDK's `URLEncoder`

Reply:
Looked at decoder, now looking more detail at encoder ;) I guess this also has the main advantage of not generating intermediate strings so we probably want to keep it.

But the Netty implementation is quite slow since it works on `CharSequence` (it probably needs `instanceof` and optimized implementations added).

Can we go ahead and do at least 1) and if not too complicated 2)?

1) Only decode when `needEncoding` character found to reduce garbage and processing

https://github.com/google/guava/blob/master/guava/src/com/google/common/net/PercentEscaper.java#L147

2) Detect non-encoded runs and use `StringBuilder.append(String, int, int)`? It looks like that would be significantly faster

Reply:
Actually - don't worry about it I started working on it for Netty codebase :)

Reply:
:laughing: OK!

Reply:
Sent https://github.com/netty/netty/pull/9887 for 1)

Reply:
Thinking of forking `PercentEscaper` to make it encode the entire query params instead of individual names/values.

Reply:
Ah yeah since we already have all params when encoding that should be the fastest of them all. I don't think we'd be able to do it in Netty - so if you feel up to it go for it :)

Reply:
I tried, but I failed to beat Guava, maybe due to how things interact with the optimizer. I always see 5-10% throughput drop whenever I modify `PercentEscaper.escape(String)` to append to a `StringBuilder` instead of returning:

```java
// Baseline:
for (Entry<String, String> e : params) {
    buf.append(guavaEscaper.escape(e.getKey()))
       .append('=')
       .append(guavaEscaper.escape(e.getValue()))
       .append('&');
}
// This is 5~10% slower:
for (Entry<String, String> e : params) {
    final String key = e.getKey();
    final String value = e.getValue();
    // isSafeOctetsOnly() is basically same with PercentEscaper.encode()
    // except that it returns true or false.
    if (isSafeOctetsOnly(key)) {
        buf.append(key);
    } else { ... }
    buf.append('=');
    if (isSafeOctetsOnly(value)) {
        buf.append(value);
    } else { ... }
    buf.append('&');
} 
```

I tried various things like:

- replace `if (isSafeOctetsOnly(...) { } else { }` with a method like:
  ```java
  static void encode(StringBuilder buf, String s) {
      for (int i = 0; i < slen; i++) {
          final char c = s.charAt(i);
          if (c >= safeOctets.length || !safeOctets[c]) {
              encodeUtf8Component(buf, s);
              return;
          }
      }

      buf.append(s);
  }

  for (Entry<String, String> e : params) {
      encode(buf, e.getKey());
      buf.append('=');
      encode(buf, e.getValue());
      buf.append('&');
  } 
  ```
- removing `else {}` completely

Reply:
So, let me replace `QueryStringEncoder` with `UrlEscapers` in this PR. Better performance and fewer code. Maybe we can revisit again later.

Reply:
What's also interesting is that the performance is still 5~10% lower when we delegate to `appendQueryString()` in `toQueryString()`. Inlining `appendQueryString()` gives an instant boost. (Note that still, we lose 5~10% again once we pass `StringBuilder` as a parameter somewhere.)

Reply:
Seems like escape analysis is acting on. Switching to thread-local reused string builder removes the performance gap, i.e. let me continue optimizing ours.

Reply:
Optimization is complete. Will post the result later.